### PR TITLE
Implement PartialEq for proc_macro::Ident == strings

### LINKED ISSUE #51074: Implement PartialEq for Ident, PartialEq for TokenStream
The original hesitation around these impls was whether to consider things equal that have different `Span`s.

I propose that if somebody is writing a macro that relies on knowing whether two tokens are the very same down to their syntax context, we should consider that an unhelpable case of "doing it wrong." On the other hand comparing the text of an Ident is quite common and perfectly correct. For example when iterating through field attributes in a custom derive to handle something like `#[serde(rename = "...")]` it makes sense to compare whether the idents in the token stream are `== i_serde` or `== i_rename`.

tracking issue: #38356
cc @Eijebong @alexcrichton

### PR DESCRIPTION
**Pull request by dtolnay**
```rust
impl PartialEq<str> for Ident
```

```rust
impl PartialEq<String> for Ident
```

```rust
impl<'a, T> PartialEq<&'a T> for Ident
where
    T: ?Sized,
    Ident: PartialEq<T>
```

Closes https://github.com/rust-lang/rust/issues/51074.

We avoid implementing PartialEq\<Ident\> because *"comparison in vacuum doesn't generally make sense in presence of hygiene and it became a source of bugs"* --- see https://github.com/rust-lang/rust/issues/51074#issuecomment-392253878. However, comparing an Ident to a string is *"probably explicit enough"* --- see https://github.com/rust-lang/rust/issues/51074#issuecomment-392388309.

The use case is to support attribute parsing, among other things. For example parsing #\[serde(rename = "...")\] might involve `if ident == "serde"` and `if ident == "rename"`.

### CODE REVIEW

**Comment by est31** [comment]
:+1: Recently I've looked for an allocation free method for comparing idents with str's and couldn't find one. Had to use `to_string` which allocates.

File: library/proc_macro/src/lib.rs [hunk 1]
```diff
@@ -923,6 +923,31 @@ impl fmt::Debug for Ident {
...
+    fn eq(&self, rhs: &String) -> bool {
+        <Self as PartialEq<str>>::eq(self, rhs.as_str())
+    }
+}
+
+#[stable(feature = "proc_macro_ident_eq", since = "1.49.0")]
+impl<'a, T> PartialEq<&'a T> for Ident
+where
+    T: ?Sized,
+    Ident: PartialEq<T>,
```
Comment (petrochenkov) [inline library/proc_macro/src/lib.rs:944]: Not sure what is the library policy regarding impls like this, in particular.

**Comment by petrochenkov** [comment]
I've nominated both this PR and the `char` PR for the libs team since they are instant stabilizations.
Both look reasonable to me.

**Comment by dtolnay** [comment]
@rfcbot fcp merge

**Comment by c410-f3r** [comment]
cc https://github.com/rust-lang/rust/pull/59070

**Comment by est31** [comment]
I thought about this a little. I wonder about raw vs non-raw idents and whether there should be something to accommodate for it. As in, should an Ident `r#hello` compare equal with `"hello"` or not.

If you are checking for keywords, then you don't want them to compare equal to each other. If you are checking for variable/function names etc, then usually you want `r#hello` be equivalent to `"hello"`. So the problem is not solved by changing defaults (making `r#hello` compare equal with `"hello"`).

In my puny little proc macro that I'm writing, I'm mainly comparing with keywords, so having them be not equal is good for me, but maybe some folks want to compare to function/variable names and they'll have to do `if id == "hello" || id == "r#hello"`.

Maybe there could be a function like `fn ignore_rawness(&self) -> IdentRawnessIgnored<'a>` which impl's eq with strings as well, but ignores the rawness? There might also be "raw keywords" in the future which always eval to keywords, making it easier to use new keywords. Then maybe a different name would be nicer. I think both can be added backwards compatibly, and I think `id == "hello"` is a good shorthand for `id.to_string() == "hello"`, so this PR is a step into the right direction. But important to keep this in mind.

**Comment by dtolnay** [comment]
We do pretty much that in Syn. There is a function `unraw` (https://docs.rs/syn/1.0.48/syn/ext/trait.IdentExt.html#tymethod.unraw) with signature &amp;Ident -\> Ident, so e.g. you'd write `ident.unraw() == "value"`.

**Comment by petrochenkov** [comment]
I've just checked and looks like the standard library doesn't currently any have any `PartialEq` impls recursively unwrapping references like https://github.com/rust-lang/rust/pull/78634#discussion_r515676575 does.

**Comment by dtolnay** [comment]
I am not dead set on that exact impl signature, but I have found some level of reference unwrapping quite key in practice specifically for Ident. proc_macro2 accomplishes the reference unwrapping by having:

```rust
impl<T> PartialEq<T> for Ident
where
    T: ?Sized + AsRef<str>;
```

where `String` and `str` obviously implement `AsRef<str>`, and then `&U` implements `AsRef<str>` whenever `U` does, unwrapping any number of layers of references via this impl:

```rust
impl<U, V> AsRef<V> for &'_ U
where
    U: ?Sized + AsRef<V>,
    V: ?Sized;
```

However I felt `AsRef<str>` was maybe too vague of a bound to use for proc_macro::Ident and preferred requiring the rhs to provide a PartialEq explicitly instead in order to compare with Ident.

File: library/proc_macro/tests/test.rs [hunk 2]
```diff
@@ -10,3 +10,14 @@ fn test_line_column_ord() {
...
 }
+
+#[test]
+fn test_ident_eq() {
+    // Good enough if it typechecks, since proc_macro::Ident can't exist in a test.
+    fn _check(ident: &Ident, string: String) {
+        let _ = ident == "serde";
+        let _ = *ident == "serde";
+        let _ = *ident == string;
+        let _ = *ident == &&&string;
```
Comment (petrochenkov) [inline library/proc_macro/tests/test.rs:21]: Could you add some tests for raw identifiers?

**Comment by m-ou-se** [comment]
> I've just checked and looks like the standard library doesn't currently any have any `PartialEq` impls recursively unwrapping references like [#78634 (comment)](https://github.com/rust-lang/rust/pull/78634#discussion_r515676575) does.

This concern was discussed in the libs team meeting of 2020-11-18. Although this hasn't been done before in `std`, the consensus was that this is fine in this `proc_macro` addition.

(Thanks for pointing it out!)

**Comment by petrochenkov** [comment]
Closing due to inactivity.

**Pull request closed without merging**